kong/clustering/version_negotiation/init.lua

```diff
@@ -0,0 +1,340 @@
+
+local cjson = require "cjson.safe"
+local pl_file = require "pl.file"
+local http = require "resty.http"
+
+local constants = require "kong.constants"
+local clustering_utils = require "kong.clustering.utils"
+
+local ngx = ngx
+local ngx_log = ngx.log
+local ngx_ERR = ngx.ERR
+local ngx_DEBUG = ngx.DEBUG
+local _log_prefix = "[version-negotiation] "
+
+local KONG_VERSION
+local CLUSTERING_SYNC_STATUS = constants.CLUSTERING_SYNC_STATUS
+
+
+local _M = {}
+
+
+local function get_body()
+  ngx.req.read_body()
+  local body = ngx.req.get_body_data()
+  if body then
+    return body
+  end
+
+  local fname = ngx.req.get_body_file()
+  if fname then
+    return pl_file.read(fname)
+  end
+
+  return ""
+end
+
+local function response(status, body)
+  ngx.status = status
+
+  if type(body) == "table" then
+    ngx.header["Content-Type"] = "application/json"
+    body = cjson.encode(body)
+  end
+
+  ngx.say(body)
+  return ngx.exit(status)
+end
+
+local function response_err(msg)
+  return response(400, { message = msg })
+end
+
+local function verify_request(body)
+  if type(body.node) ~= "table" then
+    return false, "field \"node\" must be an object."
+  end
+
+  if type(body.node.id) ~= "string" then
+    return false, "field \"node.id\" must be a string."
+  end
+
+  if type(body.node.type) ~= "string" then
+    return false, "field \"node.type\" must be a string."
+  end
+
+  if type(body.node.version) ~= "string" then
+    return false, "field \"node.version\" must be a string."
+  end
+
+  if type(body.services_requested) ~= "table" then
+    return false, "field \"services_requested\" must be an array."
+  end
+
+  return true
+end
+
+
+local function node_info()
+  return {
+    id = kong.node.get_id()
+  }
+end
+
+local function cp_priority(name, req_versions, known_versions)
+  local versions_set = {}
+  for _, version in ipairs(req_versions) do
+    versions_set[version] = true
+  end
+
+  for _, v in ipairs(known_versions) do
+    if versions_set[v.version] then
+      return true, {
+        name = name,
+        version = v.version,
+        message = v.message,
+      }
+    end
+  end
+
+  return false, { name = name, message = "No valid version" }
+end
+
+local all_known_services = require "kong.clustering.version_negotiation.services_known"
+
+local function check_node_compatibility(client_node)
+  if client_node.type ~= "KONG" then
+    return nil, ("unknown node type %q"):format(client_node.type), CLUSTERING_SYNC_STATUS.UNKNOWN
+  end
+
+  local ok, msg, status = clustering_utils.check_kong_version_compatibility(KONG_VERSION, client_node.version)
+  if not ok then
+    return ok, msg, status
+  end
+
+  return true, nil, CLUSTERING_SYNC_STATUS.NORMAL
```

Comment:
```suggestion
  return clustering_utils.check_kong_version_compatibility(KONG_VERSION, client_node.version)
```

Maybe this would do?